###### Raw extraction of clades corresponding N1 ######

# N1 genes corresponding to gsgd clades are identified with the colored HA tree 
# `S4_pH5-slc-x2-gsgd_colclade1.tre`. Three clades in China (+Hong Kong) are 
# then subjected to rmdup. iqtrees generated by rmduped sequeces are grouped 
# as `HA_subclade.R` (but with sticter criteria - within 1 year; similarity).
# Alignemnts are subsampled using functions `pre_sumsample` and 
# `subsample_gsgd` as `_clades_subsample_`. We do not removed outliers in this 
# analyses here because the diversity of NA is naturally shpaed by both point
# mutations and reassortment. However, we run BEAST with `-x2` batch which 
# are identical to `-x1` files.


require( purrr )
require( stringr )
source( "../functions/function.R" )

# 1 extract each clade ------

source("../functions/function.R")

# slc_x2_gsgd_colclade = "N1TONX/processed/pH5Nx/slc-x2/iqtree/S4_pH5-slc-x2-gsgd_colclade1.tre"
# raw_fas              = "N1TONX/processed/pH5Nx/slc-x2/S6_pN1-slc-x2.fasta"
# 
# tag_table = tagExtra( slc_x2_gsgd_colclade )
# 
# assigned_name = c( "c234N1", "c232", "c7" )
# col_code      = c( "ff6666", "ccff66", "66ccff" )
# 
# fas_id0  = fastaEx( raw_fas )$id
# fas_seq0 = fastaEx( raw_fas )$seq
# 
# for( i in 1:3 )
# {
#   idx = grep( col_code[i], tag_table$tag )
# 
#   ha_c    = tag_table$id[idx]
#   ha_id_c = str_match( ha_c, "^[0-9A-Z_]+" )[,1]
#   ha_id_c = ha_id_c[ -which( ha_id_c == "UNID" ) ]
#   
#   na_id  = str_match( fas_id0, "^[0-9A-Z_]+" )[,1]
#   
#   m = match( ha_id_c, na_id )
#   m = m[ -which(is.na(m)) ]
#   
#   outname = paste0( "N1TONX/processed/N1_subclade/raw/S6_pN1-", assigned_name[i], ".fasta" )
# 
#   print( length(m) )
#   write.fasta( fas_seq0[m], fas_id0[m], outname )
# }

# manually remove the gaps in the alignments

# 2 CN-HK ------

# source( "../functions/function.R" )
# 
# path_tsv     = "N1TONX/raw/pH5Nx/"
# path_sub_raw = "N1TONX/processed/N1_subclade/raw/"
# 
# input  = list.files( path_tsv, full.names = TRUE, pattern = "^S.*tsv$" )
# df_tsv = do.call( rbind,
#                   lapply( as.list( input ),
#                           function(x) read.table( x, header = TRUE, fill = TRUE, sep = "\t", stringsAsFactors = FALSE ) ) )
# 
# list_fasta = list.files( path_sub_raw, pattern = ".fasta", full.names = TRUE )
# 
# for( i in 1: length(list_fasta) )
# {
#   fas_seq = fastaEx( list_fasta[i] )$seq
#   fas_id  = fastaEx( list_fasta[i] )$id
# 
#   x =
#   map( fas_id,
#        function(x)
#        {
#          ac0 = str_split( x, "__")[[1]][1]
# 
#          ac = str_extract( ac0, "[A-Z0-9]+$" )
#          m  = match( ac, df_tsv$ac )
# 
#          if( is.na(m) | length(m) > 1 ){ stop("mismatch") }
# 
#          y = grepl( "China|Hong_Kong", df_tsv$country[m], ignore.case = TRUE )
#          return(y)
#        })
# 
#   z = unlist(x)
#   write.fasta( sequences = fas_seq[z], names = fas_id[z],
#                file.out  = gsub( ".fasta", "-CN.fasta", list_fasta[i]  ) )
# }

# 3 rmdup ------

source( "../functions/f_rmdup.R" )
 
# sub_CN   = list.files( "N1TONX/processed/N1_subclade/slc-CN/", pattern = "^S.*.fasta", full.names = TRUE)
# path_tsv = "N1TONX/raw/pH5Nx/S6_pNx.tsv"
# 
# for( i in 1:length(sub_CN) )
# {
#   fast_rmdup_v3( fas_file   = sub_CN[i],
#                  table_file = path_tsv )
# }


# 4 grouping and subsampling ------

# manually group as gsgd / H5_subclade dataset 

source( "_clades_subsample_.R" )

# pre_sumsample( "N1TONX/processed/N1_subclade/slc-CN/rmdup/iqtree/S6_pN1-c232-CN_rmDup_col2.tre",
#                yr_range = 1,
#                batchname = "c232")
# pre_sumsample( "N1TONX/processed/N1_subclade/slc-CN/rmdup/iqtree/S6_pN1-c234N1-CN_rmDup_col2.tre",
#                yr_range = 1,
#                batchname = "c234N1")
# pre_sumsample( "N1TONX/processed/N1_subclade/slc-CN/rmdup/iqtree/S6_pN1-c7-CN_rmDup_col2.tre",
#                yr_range = 1,
#                batchname = "c7")

# subsampling 

input_sam_table = list.files( "N1TONX/processed/N1_subclade//CN-x1",
                              pattern = "sam_table", full.names = TRUE )
input_fas       = list.files( "N1TONX/processed/N1_subclade/raw/",
                              pattern = "^S[0-9]_.*.fasta$", full.names = TRUE )

fas_id = unlist( sapply( as.list( input_fas ),
                          function(x)
                          {
                            y = fastaEx(x)$id
                          } ) )

fas_seq = lapply( as.list( input_fas ),
                         function(x)
                         {
                           y = fastaEx(x)$seq
                         } )
fas_seq = do.call( c, fas_seq )


nseed = c( 123, 234, 345, 456, 567 )

for( i in 1: length(input_sam_table) )
{
  tab = read.table( input_sam_table[i], sep = "\t", stringsAsFactors = FALSE, header = TRUE )

  for( j in 1:5 )
  {
    ran = subsample_gsgd( xseed = nseed[j], path_sam_table = input_sam_table[i] )

    m = match( tab$name[ ran ], fas_id )
    if( TRUE %in% is.na(m) ){ stop( "mismatch" ) }

    clade_i  = str_extract( input_sam_table[i], "c[0-9N]{1,6}_" )
    filename = paste0( "N1TONX/processed/N1_subclade/CN-x1//S6_pN1-", clade_i, "samX", j, ".fasta" )

    write.fasta( fas_seq[m], fas_id[m], filename )
  }
 print(i)
}




